### ISSUE TITLE
2.0rc关于dataloader数据集加载问题

### ISSUE THREAD

**Issue by besttrying** [PaddlePaddle/Paddle#29763]
在aistudio平台上训练，安装paddlepaddle2rc版本，
![AO20{GX)NP%}VDDN~$AXMQ8](https://user-images.githubusercontent.com/18502224/102600234-8b4e6e80-4159-11eb-8529-432136a1a5b7.png)
在加载图片过程中，会在加载过程中报错。

**Comment by sandyhouse** [comment]
数据读取报错，请检查下数据组织结构和数据读取代码是否一致。

**Comment by besttrying** [comment]
检查过，一致的。这个并不是完全不能读取图片，而是在读取的过程中突然报错。报错的图片路径每次都不一样，我查看后对应的路径下确实有图片。

**Comment by sandyhouse** [comment]
提供下最小可复现代码吧

**Comment by besttrying** [comment]
import os
import csv
import random
import PIL.Image as Image
import numpy as np
from paddle.io import Dataset
import glob


def get_class(sample):
    return sample.split('/')[-2]


class Reader(Dataset):
    def __init__(self, data_path, mode='train', episode_num=10000, support_class_num=5, support_shot_num=1, query_num=15, transform=None):

        super(Reader, self).__init__()
        self.data_path = data_path
        self.mode = mode
        self.episode_num = episode_num
        self.support_class_num = support_class_num
        self.support_shot_num = support_shot_num
        self.query_num = query_num
        self.transform = transform
        self.data = self.reader(self.data_path, self.episode_num, self.support_class_num, self.support_shot_num, self.query_num)

    def reader(self, data_path, episode_num, support_class_num, support_shot_num, query_num):
        class_list = glob.glob(data_path + '/n*')
        class_list = [os.path.join(data_path, c) for c in class_list]
        episode = []
        for i in range(episode_num):
            class_folders = random.sample(class_list, self.support_class_num)
            labels = np.array(range(len(class_folders)))
            labels = dict(zip([f.split('/')[-1] for f in class_folders], labels))
            samples = dict()

            
            sample_roots = []
            query_roots = []
            for c in class_folders:
                temp = [os.path.join(c, x) for x in os.listdir(c)]
                samples[c] = random.sample(temp, len(temp))
                random.shuffle(samples[c])

                sample_roots += samples[c][:self.support_shot_num]
                query_roots += samples[c][self.support_shot_num:self.support_shot_num + self.query_num]
            np.random.shuffle(query_roots)
            sample_labels = [labels[get_class(x)] for x in sample_roots]
            # print('sample_labels len ', len(sample_labels))
            # print('sample_labels ', sample_labels)
            query_labels = [labels[get_class(x)] for x in query_roots]
            # print('query_labels len ', len(query_labels))
            # print('query_labels ', query_labels)
            data_file = {
                'support_images': sample_roots,
                'supoort_labels': sample_labels,
                'query_images': query_roots,
                'query_labels': query_labels
            }
            episode.append(data_file)
        return episode
    def __len__(self):
        return len(self.data)

    def __getitem__(self, index):
        

        # np.random.shuffle(sample_roots)
        data = self.data[index]
        sample_roots = data['support_images']
        query_roots = data['query_images']

        sample_labels = data['supoort_labels']
        query_labels = data['query_labels']
        sample_img = []
        query_img = []
        # print('ssample_roots ', sample_roots)
        for im_file in sample_roots:
            # print('sqmple:', im_file)
            img = Image.open(im_file)
            # print('img')
            if img.mode != 'RGB':
                img = img.convert('RGB')

            if self.transform is not None:
                img = self.transform(img)
            sample_img.append(img)

        # print('query_roots ', len(query_roots))
        i = 0
        for im_file in query_roots:
            i += 1
            # print('query:', im_file)
            img = Image.open(im_file)
            # print('img', i)
            if img.mode != 'RGB':
                img = img.convert('RGB')

            if self.transform is not None:
                img = self.transform(img)
            query_img.append(img)
        yield (sample_img, sample_labels, query_img, query_labels)


    train_data = Reader(arg.data_path, transform=transform)
    train_data = DataLoader(train_data, batch_size=1, shuffle=True, num_workers=0)
    # val_data = Reader(arg.data_path, mode='val')

    for epoch, (support_images, support_targets, query_images, query_targets) in enumerate(train_data):

    目前能读取图片，然而在这个循环阶段报同样的错

**Comment by sandyhouse** [comment]
报错内容是什么

**Comment by besttrying** [comment]
> 报错内容是什么

![%F@ %ZAXZ5NC%W$ZDG Y~W6](https://user-images.githubusercontent.com/18502224/102616559-b09ba680-4172-11eb-82a4-c79ed4cfd7fb.png)

**Comment by sandyhouse** [comment]
> 目前能读取图片，然而在这个循环阶段报同样的错

这不还是reader部分报错吗，我还以为是其它地方报错了呢。请提供下完整的最小可复现代码和数据集，及数据组织格式和启动脚本。我们本地复现下看看。

**Comment by besttrying** [comment]
[fsl.zip](https://github.com/PaddlePaddle/Paddle/files/5715777/fsl.zip)

数据集为mini-imagenet，这里无法上传。通过处理train.csv等，把图片分为train，test，val三个文件夹。

**Comment by besttrying** [comment]
[miniImageNet.zip](https://github.com/PaddlePaddle/Paddle/files/5715815/miniImageNet.zip)
这是三个csv文件

**Comment by sandyhouse** [comment]
> 通过处理train.csv等，把图片分为train，test，val三个文件夹。`

也烦请提供下处理脚本，以及训练的运行命令吧。

**Comment by besttrying** [comment]
图片处理文件在主文件夹中的/data/proc_image.py
训练命令在主文件夹下python tools/train.py即可。
麻烦了！非常感谢

**Comment by sandyhouse** [comment]
> 数据集为mini-imagenet，这里无法上传。

这个数据集哪里可以下载到吗，可以给个链接。我没有这个数据集。

**Comment by besttrying** [comment]
> > 数据集为mini-imagenet，这里无法上传。
> 
> 这个数据集哪里可以下载到吗，可以给个链接。我没有这个数据集。

百度云链接: https://pan.baidu.com/s/1npRhZajLrLe6-KtSbJsa1A 密码: ztp5
我目前只有百度云的源

**Comment by sandyhouse** [comment]
数据集有点大，百度云下载比较慢。请问我是否可以用ImageNet数据集模拟，数据组织格式是什么样的？

**Comment by besttrying** [comment]
可以用imagenet模拟，mini-imagenet中有四个文件，一个images文件夹包括了所有的图像，然后根据三个csv文件把images里面的图像取出来存放到train，test，val文件夹下。以train文件夹为例，里面包括所有的训练图片，而每一类的训练图片又放在同类别的文件夹下面。相当于imagenet中每个类别的压缩包解压到train文件下对应的类别文件即可。imagenet中按照下面建好文件夹，把对应的图片解压放进去就行。图片创建好后，在tools/train.py中args中修改图片路径即可。
paddle新手，这里真是劳烦了。着实感谢。
若是麻烦，实在不行就不用了。我再研究研究，可能是我在加载数据的时候哪个小地方没注意导致的。
mini-imagenet
 --------images
 ----------image1
  ---------........
  -------train
  ---------class1
   ----------image1
----------............
 ----------class2
 -----------image1
 -----------.............

**Comment by sandyhouse** [comment]
几个修改点：
1. for epoch, (support_images, support_targets, query_images, query_targets) in enumerate(train_data): --> for epoch, (support_images, support_targets, query_images, query_targets) in enumerate(train_data()):
2. sample_img.append(img) --> sample_img.append(np.array(img))
3.  query_img.append(img) --> query_img.append(np.array(img))
4.  return (sample_img, sample_labels, query_img, query_labels) --> return (np.array(sample_img), np.array(sample_labels), np.array(query_img), np.array(query_labels))

另外，你组网上可能也有问题，再仔细排查下吧。

**Comment by besttrying** [comment]
好的！为带来的麻烦表示感谢！

**Comment by besttrying** [comment]
> 几个修改点：
> 
> 1. for epoch, (support_images, support_targets, query_images, query_targets) in enumerate(train_data): --> for epoch, (support_images, support_targets, query_images, query_targets) in enumerate(train_data()):
> 2. sample_img.append(img) --> sample_img.append(np.array(img))
> 3. query_img.append(img) --> query_img.append(np.array(img))
> 4. return (sample_img, sample_labels, query_img, query_labels) --> return (np.array(sample_img), np.array(sample_labels), np.array(query_img), np.array(query_labels))
> 
> 另外，你组网上可能也有问题，再仔细排查下吧。

非常感谢！为带来的麻烦表示感谢！
